monai/data/image_writer.py

```diff
@@ -0,0 +1,725 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Dict, Mapping, Optional, Sequence, Union
+
+import numpy as np
+import torch
+
+from monai.apps.utils import get_logger
+from monai.config import DtypeLike, NdarrayOrTensor, PathLike
+from monai.data.utils import compute_shape_offset, create_file_basename, ensure_mat44, to_affine_nd
+from monai.networks.layers import AffineTransform
+from monai.transforms.spatial.array import Resize
+from monai.transforms.utils_pytorch_numpy_unification import ascontiguousarray, moveaxis
+from monai.utils import (
+    GridSampleMode,
+    GridSamplePadMode,
+    InterpolateMode,
+    OptionalImportError,
+    convert_data_type,
+    ensure_tuple,
+    ensure_tuple_rep,
+    look_up_option,
+    optional_import,
+    require_pkg,
+)
+
+AFFINE_TOL = 1e-3
+DEFAULT_FMT = "%(asctime)s %(levelname)s %(filename)s:%(lineno)d - %(message)s"
+logger = get_logger(module_name=__name__, fmt=DEFAULT_FMT)
+
+if TYPE_CHECKING:
+    import itk  # type: ignore
+    import nibabel as nib
+    from nibabel.nifti1 import Nifti1Image
+    from PIL import Image as PILImage
+
+    has_itk = has_nib = has_pil = True
+else:
+    itk, has_itk = optional_import("itk", allow_namespace_pkg=True)
+    nib, has_nib = optional_import("nibabel")
+    Nifti1Image, _ = optional_import("nibabel.nifti1", name="Nifti1Image")
+    PILImage, has_pil = optional_import("PIL.Image")
+
+__all__ = [
+    "ImageWriter",
+    "ITKWriter",
+    "NibabelWriter",
+    "PILWriter",
+    "FolderLayout",
+    "SUPPORTED_WRITERS",
+    "register_writer",
+    "resolve_writer",
+]
+
+SUPPORTED_WRITERS: Dict = {}
+
+
+def register_writer(data_format, *im_writer):
+    """
+    Register ``ImageWriter``, so that writing a file of ``data_format``
```

Comment:
Format is mentioned in a number of places but we don't specify what we mean by format. Is this meant to be the file extension or some identifier of a format? For common files like "png" it would be both the extension and the name of the format, but for JPEG we have "jpg" or "jpeg" as valid extensions so which would be use as the format name needs to be resolved. Same with NIfTI, the extension could be "nii" or "nii.gz".

This raises the question if we want to support any file compressed with gzip adding ".gz" to the extension, a file with "png" or "png.gz" should resolve to the same writers.

Reply:
thanks, for now I updated to have the mapping "filename extension" <> imagewriter,  we can have an additional layer such as: "filename extension" <> fileformat <> imagewriter the latest is at https://github.com/Project-MONAI/MONAI/pull/3773.